### PR TITLE
memory: switch to the latest tcmalloc for x86_64 builds

### LINKED ISSUE #10053: allow new tcmalloc to be used as an option
https://abseil.io/blog/20200212-tcmalloc

Currently only supported on linux/x86_64 so we could add as a build option or potentially just rip out the old one and everyone else would get standard malloc (sub-optimal given the rise of ARM).

### PR DESCRIPTION
**Pull request by rojkov**
Commit Message: memory: switch to the latest tcmalloc for x86_64 builds
Additional Description: Switch to the new tcmalloc for linux x86_64 builds. All other builds still get the old tcmalloc from gperftools. Switching profiling or the debugalloc feature on will result in linking to the old tcmalloc too. The old tcmalloc can still be enabled for x86_64 builds with `--define tcmalloc=gperftools`.
Risk Level: High
Testing: run unit tests
Docs Changes: added a note about the new `--define tcmalloc=gperftools` option to `bazel/README.md`.
Release Notes: updated the "Minor Behavior Changes" section of `version_hstory/current.rst`
Fixes #10053

### CODE REVIEW

File: bazel/repository_locations.bzl
Comment (moderation) [inline bazel/repository_locations.bzl:211]: Add date of https://github.com/google/tcmalloc/commit/d1311bf409db47c3441d3de6ea07d768c6551dec commit under this line
`#2020-09-16`

File: bazel/repository_locations.bzl
Comment (lizan) [inline bazel/repository_locations.bzl:216]: should be `dataplane`

File: source/common/memory/utils.cc
Comment (lizan) [inline source/common/memory/utils.cc:18]: Move this into `releaseFreeMemory`?

Comment (rojkov) [reply source/common/memory/utils.cc:18]: It's used in two functions: `releaseFreeMemory()` and `tryShrinkHeap()`.

File: bazel/repository_locations.bzl
Comment (rojkov) [reply bazel/repository_locations.bzl:211]: Added the date

File: bazel/repository_locations.bzl
Comment (rojkov) [reply bazel/repository_locations.bzl]: Fixed

**Comment by rojkov** [comment]
Hm.. the way how `generic.current_allocated_bytes` is calculated in the new tcmalloc has changed a bit (per-cpu cache is substructed now). I need more tests to run -> converting to Draft for now.

**Comment by rojkov** [comment]
@jmarantz With the new tcmalloc the amount of consumed memory becomes non deterministic and from run to run I see different results of `memory_test.consumedBytes()`  in the same test. Can we relax the tests a bit and avoid the strict equivalence checks?

File: test/integration/stats_integration_test.cc
Comment (jmarantz) [inline test/integration/stats_integration_test.cc:314]: this whole test is now gone, once you merge, so you can just keep the edit for the real symbol-tables test below, and drop this one.

**Review by jmarantz — COMMENTED** [review]
> @jmarantz With the new tcmalloc the amount of consumed memory becomes non deterministic and from run to run I see different results of memory_test.consumedBytes() in the same test. Can we relax the tests a bit and avoid the strict equivalence checks?

Are you sure that deterministic byte-counts are not still available via the API in some way? It looks like maybe you changed the memory-stats APIs to include unmapped bytes?

File: source/common/memory/stats.cc
Comment (jmarantz) [inline source/common/memory/stats.cc:15]: maybe for deterministic memory checks we can just use generic.current_allocated_bytes and not include tcmalloc.cpu_free?

We can make 2 APIs if that helps.

I'm also confused by the github UI; it is not showing me the prior impl of this method.

Comment (lizan) [reply source/common/memory/stats.cc:15]: This part (from `#if` to `#elif`) is new for tcmalloc. The previous impl is below.

**Review by lizan — COMMENTED** [review]
LGTM, thanks. Defer to @jmarantz for the deterministic memory checks and then merge.

File: source/common/memory/stats.cc
Comment (rojkov) [reply source/common/memory/stats.cc:15]: > maybe for deterministic memory checks we can just use generic.current_allocated_bytes and not include tcmalloc.cpu_free?

Right, my bad. It shouldn't be included. I'll drop it. Though after the last merge all tests passed for  https://github.com/envoyproxy/envoy/pull/13251/commits/a6d4c424ce5c5c2c01c37557632e6481f028ac4c. But the version without tcmalloc.cpu_free passes the release tests too on my local host. Let's see...

> I'm also confused by the github UI; it is not showing me the prior impl of this method.

The old version is still there, just wrapped in `#if defined(GPERFTOOLS_TCMALLOC)`. It's needed for non-x86 builds.

**Comment by rojkov** [comment]
> Are you sure that deterministic byte-counts are not still available via the API in some way? It looks like maybe you changed the memory-stats APIs to include unmapped bytes?

I did include unmapped bytes to `Stats::totalCurrentlyReserved()` only. The reason is that the new tcmalloc doesn't include them now for `generic.heap_size`. See https://github.com/google/tcmalloc/blob/43cb87553037d858d819437e294bd6a4eb0210ae/tcmalloc/tcmalloc.cc#L877 vs https://github.com/gperftools/gperftools/blob/bda3c82e11615ca9e7751d1f3cfb161026ee742a/src/tcmalloc.cc#L717

File: test/common/stats/symbol_table_impl_test.cc
Comment (jmarantz) [inline test/common/stats/symbol_table_impl_test.cc:741]: CI fails on gcc now with:

Expected: (string_mem_used) <= (7775872), actual: 7792256 vs 7775872

I'm guessing this may be related to varying STL versions for std::string across platforms than tcmalloc. I think we can just remove the exact check for string_mem_used. We can keep the test that symbol tables cut the usage by 2/3.

**Review by jmarantz — APPROVED** [review]

**Review by mattklein123 — APPROVED** [review]
Awesome, thanks!

**Comment by alyssawilk** [comment]
This morning my default build was using clang+llvm-8.0.0 and build failed with "error: 'asm goto' constructs are not supported yet".  I updated clang to 9 and I'm good, but based on that and an unsubstantiated google search (http://llvm.1065342.n5.nabble.com/llvm-dev-is-clang9-supporting-asm-goto-td128867.html) I wonder if we need to update the build requirements doc (https://www.envoyproxy.io/docs/envoy/latest/install/building#requirements)

**Comment by alyssawilk** [comment]
for the record, I just restored clang-8 as default and verified it builds with --define  tcmalloc=gperftools so ideally we can just update requirements doc to note folks should use that option with older versions of clang

**Comment by rojkov** [comment]
> for the record, I just restored clang-8 as default and verified it builds with --define tcmalloc=gperftools so ideally we can just update requirements doc to note folks should use that option with older versions of clang

Right. Thanks for letting know! I'll make a PR to mention it.

Basically for non-x86 builds the current requirement is still sufficient too. Those who build Envoy for embedded very often have to deal with quite outdated toolchains from BSPs they can get from board vendors.

**Comment by alyssawilk** [comment]
I added some comments in https://github.com/envoyproxy/envoy/pull/13438 as we're hoping to cut the release today or tomorrow.
LMK what you think?

**Comment by rojkov** [comment]
Great you corrected it already! I commented the PR.

**Comment by rgs1** [comment]
Hmm this change is triggering this for our build:

```
14:21:45 warning: unknown warning option '-Wno-attribute-alias'; did you mean '-Wno-attributes'? [-Wunknown-warning-option]
14:21:45 In file included from external/com_github_google_tcmalloc/tcmalloc/static_vars.cc:26:
14:21:45 In file included from external/com_github_google_tcmalloc/tcmalloc/cpu_cache.h:30:
14:21:45 external/com_github_google_tcmalloc/tcmalloc/internal/percpu_tcmalloc.h:255:7: error: 'asm goto' constructs are not supported yet
14:21:45   asm goto(
14:21:45       ^
14:21:45 external/com_github_google_tcmalloc/tcmalloc/internal/percpu_tcmalloc.h:459:25: error: invalid output constraint '=@ccbe' in asm
14:21:45             [underflow] "=@ccbe"(underflow),
```

**Comment by rgs1** [comment]
nvm just saw the prev comment

**Pull request merged**
